### ISSUE TITLE
Renovate doesn't assign PR to anyone if one of the assignees isn't project member

### ISSUE THREAD

**Issue by kovasaurus** [renovatebot/renovate#17709]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

32.192

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### If you're self-hosting Renovate, tell us what version of the platform you run.

14.10.3-ee

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

While implementing the self-hosted renovate bot on a self-hosted GitLab instance, I tried to provide multiple assignees to the bot's general configuration. After that, I provided a single repository for general configuration and ran the bots pipeline so it could create onboarding MR to that repository. The result was that MR was created but without a single assignee, even though I could see the (DEBUG) logs from the bot saying that all users are assigned to the MR. Later I realized that not all members provided to assignees were actually part of the given project, hence it failed.

Expected behavior (for me at least) would be to assign the users eligible to be MR assignees (project members), while discarding the rest. It could also be useful to show that information inside the job logs.

Have a similar bot configuration:
`
{
   assignees: [
           'user1',
           'user2'
           'user3'
   ], 
   repositories: [
           'group/repository1'
   ],
}
`
Repro steps:
1) Make sure the 'user1' and 'user2' are part of a project under 'group/repository1'
2) Make sure the 'user3' is not part of a project under 'group/repository1'
3) Run the bot
4) Check the created MR, it will have 0 assignees (and job logs explicitly mentions that all 3 users are added as assignees)
5) Delete created MR
6) Either remove 'user3' from assignees OR add 'user3' to project 'group/repository1'
7) Run the bot again
8) Check the created MR, now including all "valid" assignees 

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Adding assignees 'user1, user2, user3' to #525 (repository=group/repository1)
DEBUG: Added assignees (repository=group/repository1)
       "assignees": ["user1", "user2", "user3"]
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

**Comment by rarkins** [comment]
Here is the code for addAssignees in GitLab: https://github.com/renovatebot/renovate/blob/7232f7149d902ef64b6da62b13b358afcd675f27/lib/modules/platform/gitlab/index.ts#L963-L983

It would seem that the add function isn't throwing, otherwise you'd get a warning message.

Before we can progress, someone needs to manually test the GitLab API using the above steps to confirm that it returns 200 OK even if adding assignees fails.

**Comment by kovasaurus** [comment]
I can confirm that GitLab API returns 200 in case an assignee is not part of the project.
User id 34 Is part of the project, while user id 62 is not.
After a PUT request, they return the entire object unchanged in case of the non-eligible user for the assignee. 

![image](https://user-images.githubusercontent.com/28956838/189291928-7ad82aa0-6c94-42e6-8117-fde8a6734383.png)

![image](https://user-images.githubusercontent.com/28956838/189290359-ec3e5c05-e7fa-48e7-a4af-7d32fc4edaea.png)

Afterwards, if I provide the correct users to the request, it adds them as assignees.

![image](https://user-images.githubusercontent.com/28956838/189291647-ce58f106-7a7c-4efe-b21c-2dc753aae290.png)

![image](https://user-images.githubusercontent.com/28956838/189291115-006c9285-be4a-47ac-8aae-3e3d8819e8ee.png)

**Comment by viceice** [comment]
So there is no easy way to detect the invalid user from that request. and because it succeeds with 200, we don't detect any error. So I'm not sure we can fix this.

**Comment by kovasaurus** [comment]
There is a way to "handle" this with 1 additional API call before updating MR. 
GitLab provides a way to fetch project members, which can be filtered out by user ids.

![image](https://user-images.githubusercontent.com/28956838/189297156-e5f9c88f-ee5c-4115-844d-7d3291f67a75.png)

That way, we can use the array of id's created in for loop (lines 970-972) and query it over the mentioned endpoint. Then we can use returned array from GitLab as IDs to assign to MR. Since we have queried with specific IDs, we expect an array of the same length. If fewer items are inside, the user can get a warning, and if the size is 0 error can be thrown.

If that solution is viable for you, I can submit the PR later today.

**Comment by rarkins** [comment]
Thanks for your research! Can that endpoint be queried by a token with only Developer-level permissions, or does it require someone with a higher level of permissions?

**Comment by kovasaurus** [comment]
Yes, Developer-level permission is enough.

**Comment by rarkins** [comment]
Sounds good, but can we do this as a *fallback* instead of with every MR? e.g. something like this:
- Attempt to add assignees if configured
- If the returned list is empty, and only one assignee configured, log a warning and return
- If the returned list is empty, and more than one assignee, use your new trick to determine which aren't members. Warn if none are, otherwise debug (no need to throw)

**Comment by kovasaurus** [comment]
So if I understood correctly, You're proposing to check the response of 1st update MR request, compare its assignees' property to the array of id(s) sent into the request, and then handle it accordingly? 
I was thinking about that approach but then went back to the one I proposed for a few reasons:
- we have 2 additional API calls in case assignees' returned array is empty (1 to check project members and 2nd to do MR update again with valid members)
- even if one assignee is configured, it can happen that it's not a project member
- by duplicating the update MR request, we recognize that 1st one might fail (our action will fail, even though the request returns 200), making the function behave reactively. And if it checks for project members before updating MR, the function is proactive and eliminates the possibility of erroneous action.

I agree that there is no need to throw.

**Comment by rarkins** [comment]
What I'm proposing is:

- 1 API request for the happy path where all assignees are ok
- 1 API request for the sad path where there's one assignee and they fail
- 2 API requests for the sad path where there's more than one assignee and they all fail
- 3 API requests for the sad path where there's more than one assignee and not all fail

What I think you're proposing is always 2 API requests whenever there's one or more assignees.

I think it's more important to optimize for the happy path which should be the majority of the time.

**Comment by kovasaurus** [comment]
You're right, this wouldn't be a regular case. 
Just a notice, I don't think 4th bullet point is possible at this moment, they are either all assigned, or no one is.
But it is good to have it prepared for time if/when GitLab changes its route, so it handles the assignees which are not part of the project.

**Comment by rarkins** [comment]
I mean that if all the configured assignees are missing then you don't need to do any further call. But if for example 1/2 are valid you do.
